clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java

```diff
@@ -0,0 +1,238 @@
+import org.apache.kafka.common.errors.GroupAuthorizationException;
+import org.apache.kafka.common.errors.RetriableException;
+import org.apache.kafka.common.message.FindCoordinatorRequestData;
+import org.apache.kafka.common.message.FindCoordinatorResponseData;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.FindCoordinatorRequest;
+import org.apache.kafka.common.requests.FindCoordinatorResponse;
+import org.apache.kafka.common.utils.ExponentialBackoff;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.slf4j.Logger;
+
+import java.util.List;
+import java.util.Optional;
+
+public class CoordinatorManager {
+    final static int RECONNECT_BACKOFF_EXP_BASE = 2;
+    final static double RECONNECT_BACKOFF_JITTER = 0.0;
+
+    private final Logger log;
+    private final Time time;
+    private final long requestTimeoutMs;
+    private final ErrorEventHandler errorHandler;
+    private final ExponentialBackoff exponentialBackoff;
+    private final long rebalanceTimeoutMs;
+    private final Optional<String> groupId;
+
+    private CoordinatorRequestState coordinatorRequestState = new CoordinatorRequestState();
+    private long lastTimeOfConnectionMs = -1L; // starting logging a warning only after unable to connect for a while
+    private Node coordinator;
+
+
+    public CoordinatorManager(final Time time,
+                              final LogContext logContext,
+                              final ConsumerConfig config,
+                              final ErrorEventHandler errorHandler,
+                              final Optional<String> groupId,
+                              final long rebalanceTimeoutMs) {
+        this.time = time;
+        this.log = logContext.logger(this.getClass());
+        this.errorHandler = errorHandler;
+        this.exponentialBackoff = new ExponentialBackoff(
+                config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                RECONNECT_BACKOFF_EXP_BASE,
+                config.getLong(CommonClientConfigs.RECONNECT_BACKOFF_MAX_MS_CONFIG),
+                RECONNECT_BACKOFF_JITTER);
+        this.groupId = groupId;
+        this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+        this.requestTimeoutMs = config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+    }
+
+    /**
+     * Returns a non-empty UnsentRequest we need send a FindCoordinatorRequest. These conditions are:
+     * 1. The request has not been sent
+     * 2. If the previous request failed, and the retryBackoff has expired
+     * @return Optional UnsentRequest.  Empty if we are not allowed to send a request.
+     */
+    public Optional<NetworkClientDelegate.UnsentRequest> tryFindCoordinator() {
+        if (coordinatorRequestState.lastSentMs == -1) {
+            // no request has been sent
+            return makeFindCoordinatorRequest();
+        }
+
+        if (coordinatorRequestState.lastReceivedMs == -1 ||
```

Comment:
I think we can move this check to `CoordinatorRequestState` and give it a nice name. For example, `isRequestInflight`. 

Or perhaps we could even combine all of these checks into a single check? Maybe `canSendRequest`? Then we could structure this as the following:

```java
public Optional<NetworkClientDelegate.UnsentRequest> maybeFindCoordinator() {
  if (coordinator != null) {
    return Optional.empty();
  }

  if (coordinatorRequestState.canSendRequest()) {
    return Optional.of(makeFindCoordinatorRequest());
  } 

  return Optional.empty();
}
```